File: src/chatty/catalog.py
from __future__ import annotations

from pathlib import Path

from chatty import config
from chatty.database import Database
from chatty.models import (
    AgentDraft,
    KnowledgeHit,
    MarketingStrategy,
    Product,
    RecommendationRequest,
    RecommendedProduct,
    UserContext,
    UserProfile,
)
from chatty.repositories import CommerceRepository, KnowledgeRetriever


class CatalogError(RuntimeError):
    pass


class Catalog:
    """集中商品搜索与最终业务规则，避免 Agent 和 Tool 直接处理 SQL。

    数据库、仓储和检索器都是实现零件，不在接口里：调用方要的是「搜商品」「查库存」
    「检索知识」，不是「拿到连接自己写 SQL」。想直接测某个零件就直接构造那个零件，
    别从这里穿过去——测试和调用方走同一个 seam，这条对两边都成立。
    """

    def __init__(
        self,
        data_dir: str | Path | None = None,
        *,
        database_path: str | Path | None = None,
    ) -> None:
        self._data_dir = Path(data_dir or config.DATA_DIR)
        self._database = Database(database_path, data_dir=self._data_dir)
        self._repository = CommerceRepository(self._database)
        self._retriever = KnowledgeRetriever(self._database, self._data_dir)

        # 画像与排序维度使用启动投影；finalize 会重读价格和库存等响应真值。
        self.products = self._repository.list_products()
        self.profiles = self._repository.profiles()
        self.forbidden_words = self._repository.forbidden_words()
        self._templates = self._repository.marketing_strategies(self.forbidden_words)
        # 分群齐全性由 seed_database 校验：它在 Database.__init__ 里先跑，
        # 不齐就抛 SeedDataError，轮不到这里再查一遍。
        self.categories = sorted({product.category for product in self.products})

    def close(self) -> None:
        self._database.close()

    def user_profile(self, user_id: str, overrides: UserContext) -> UserProfile:
        base = self.profiles.get(
            user_id,
            UserProfile(
                user_id=user_id,
                segment="new_user",
                preferred_categories=[],
                min_price_cents=0,
                max_price_cents=1_000_000,
            ),
        )
        return base.model_copy(
            update={
                "preferred_categories": (
                    overrides.preferred_categories or base.preferred_categories
                ),
                "min_price_cents": (
                    overrides.min_price_cents
                    if overrides.min_price_cents is not None
                    else base.min_price_cents
                ),
                "max_price_cents": (
                    overrides.max_price_cents
                    if overrides.max_price_cents is not None
                    else base.max_price_cents
                ),
                "recent_views": overrides.recent_views or base.recent_views,
                "recent_purchases": overrides.recent_purchases or base.recent_purchases,
            }
        )

    def search(
        self,
        *,
        profile: UserProfile,
        categories: list[str],
        min_price_cents: int,
        max_price_cents: int,
        tags: list[str],
        limit: int,
    ) -> list[Product]:
        category_filter = {value.casefold() for value in categories if value.strip()}
        if min_price_cents < 0 or max_price_cents <= 0 or min_price_cents > max_price_cents:
            raise CatalogError("invalid_product_search_price_range")
        if not 1 <= limit <= 20:
            raise CatalogError("invalid_product_search_limit")
        # 传了类目但全是空白 → 会退化成不过滤，模型以为筛了实际没筛，宁可报错
        if categories and not category_filter:
            raise CatalogError("invalid_product_search_categories")
        tag_filter = {value.casefold() for value in tags if value.strip()}
        if tags and not tag_filter:
            raise CatalogError("invalid_product_search_tags")
        # Tool 参数负责召回；画像价格范围仍会在 finalize 再次校验。
        candidates = [
            product
            for product in self.products
            if min_price_cents <= product.price_cents <= max_price_cents
            and (not category_filter or product.category.casefold() in category_filter)
            and (not tag_filter or tag_filter.intersection(tag.casefold() for tag in product.tags))
        ]
        return sorted(
            candidates,
            key=lambda product: self._score(product, profile),
            reverse=True,
        )[:limit]

    def _score(self, product: Product, profile: UserProfile) -> float:
        """给商品打分：热度打底，再叠加画像匹配、近期行为和价格区间三个信号。"""
        preferred = {value.casefold() for value in profile.preferred_categories}
        signals = {value.casefold() for value in profile.recent_views + profile.recent_purchases}
        searchable = {
            product.name.casefold(),
            product.category.casefold(),
            *(tag.casefold() for tag in product.tags),
        }
        score = product.popularity_score * 0.55
        if product.category.casefold() in preferred:
            score += 0.25
        if signals.intersection(searchable):
            score += 0.15
        if profile.min_price_cents <= product.price_cents <= profile.max_price_cents:
            score += 0.05
        return round(min(score, 1.0), 4)

    def inventory(self, product_ids: list[str]) -> list[Product]:
        return self._repository.inventory(product_ids)

    def retrieve_knowledge(
        self,
        query: str,
        *,
        categories: list[str],
        product_ids: list[str],
        limit: int,
    ) -> list[KnowledgeHit]:
        if not 1 <= limit <= 8:
            raise CatalogError("invalid_knowledge_limit")
        return self._retriever.retrieve(
            query,
            categories=categories,
            product_ids=product_ids,
            limit=limit,
        )

    def marketing_strategy(self, segment: str) -> MarketingStrategy:
        if segment not in self._templates:
            raise CatalogError("unknown_marketing_segment")
        return self._templates[segment]

    def finalize(
        self,
        draft: AgentDraft,
        request: RecommendationRequest,
        profile: UserProfile,
    ) -> list[RecommendedProduct]:
        # 重查 SQLite：模型可能记错价格，且从查库存到现在库存也可能变了
        current_products = {
            product.product_id: product for product in self._repository.list_products()
        }
        recommendations: list[RecommendedProduct] = []
        seen: set[str] = set()  # 防止模型把同一个商品推荐两次
        for item in draft.recommendations:
            if item.product_id in seen:
                continue
            product = current_products.get(item.product_id)
            # 商品在库里根本不存在 → 说明模型编了一个 ID，这是**异常**，直接失败。
            if product is None:
                raise CatalogError("unknown_recommended_product")
            seen.add(item.product_id)
            # 售罄或超预算是正常的业务变化，跳过即可（上面 raise 是模型错了）
            if (
                product.stock <= 0
                or product.price_cents < profile.min_price_cents
                or product.price_cents > profile.max_price_cents
            ):
                continue
            recommendations.append(
                RecommendedProduct(
                    # 下面这些业务字段**全部取自数据库**，不采信模型说的任何数值。
                    product_id=product.product_id,
                    name=product.name,
                    category=product.category,
                    price_cents=product.price_cents,
                    brand=product.brand,
                    stock=product.stock,
                    tags=product.tags,
                    score=self._score(product, profile),
                    low_stock=product.stock <= 100,
                    # 只有这两项来自模型，而且要先过一遍禁词替换。
                    reason=self._sanitize(item.reason),
                    marketing_copy=self._sanitize(item.marketing_copy),
                )
            )
            if len(recommendations) >= request.num_items:
                break  # 够数就停，多的丢掉
        # 全被过滤光了，宁可报错也不返回空列表冒充成功。
        if not recommendations:
            raise CatalogError("no_available_recommendations")
        return recommendations

    def _sanitize(self, text: str) -> str:
        """把营销禁词替换成 ***（比如"打折""优惠券"这类合规敏感词）。"""
        for word in self.forbidden_words:
            text = text.replace(word, "***")
        return text

File: apps/api/src/catalog.ts
import { readFileSync } from "node:fs";
import { join } from "node:path";

import { DATA_DIR } from "./config.js";
import { Database, segmentForIndex } from "./database.js";
import type {
  KnowledgeHit,
  MarketingStrategy,
  Product,
  RecommendationDraftItem,
  RecommendationRequest,
  RecommendedProduct,
  UserContext,
  UserProfile,
  UserSegment,
} from "./types.js";

export class CatalogError extends Error {}

type SqlRow = Record<string, unknown>;

const productFromRow = (row: SqlRow): Product => ({
  product_id: String(row.product_id),
  name: String(row.name),
  category: String(row.category),
  price_cents: Number(row.price_cents),
  description: String(row.description),
  brand: String(row.brand),
  seller_id: String(row.seller_id),
  stock: Number(row.stock),
  tags: JSON.parse(String(row.tags_json)) as string[],
  popularity_score: Number(row.popularity_score),
  image_url: String(row.image_url),
  source: String(row.source),
});

const profileFromRow = (row: SqlRow): UserProfile => ({
  user_id: String(row.user_id),
  segment: String(row.segment) as UserSegment,
  preferred_categories: JSON.parse(
    String(row.preferred_categories_json),
  ) as string[],
  min_price_cents: Number(row.min_price_cents),
  max_price_cents: Number(row.max_price_cents),
  recent_views: JSON.parse(String(row.recent_views_json)) as string[],
  recent_purchases: JSON.parse(String(row.recent_purchases_json)) as string[],
});

const normalize = (values: string[]): Set<string> =>
  new Set(
    values
      .filter((value) => value.trim())
      .map((value) => value.toLocaleLowerCase()),
  );

export class Catalog {
  readonly products: Product[];
  readonly profiles: Map<string, UserProfile>;
  readonly forbiddenWords: string[];
  readonly categories: string[];
  readonly #database: Database;
  readonly #templates: Map<string, MarketingStrategy>;
  readonly #synonyms: Map<string, string[]>;

  constructor(options: { databasePath?: string; dataDir?: string } = {}) {
    const dataDir = options.dataDir ?? DATA_DIR;
    this.#database = new Database(options.databasePath, dataDir);
    this.products = this.#listProducts();
    this.profiles = new Map(
      (
        this.#database.connection
          .prepare("SELECT * FROM user_profiles ORDER BY user_id")
          .all() as SqlRow[]
      ).map((row) => {
        const profile = profileFromRow(row);
        return [profile.user_id, profile];
      }),
    );
    this.forbiddenWords = (
      this.#database.connection
        .prepare("SELECT word FROM forbidden_words ORDER BY rowid")
        .all() as SqlRow[]
    ).map((row) => String(row.word));
    this.#templates = new Map(
      (
        this.#database.connection
          .prepare("SELECT * FROM marketing_templates")
          .all() as SqlRow[]
      ).map((row) => [
        String(row.segment),
        {
          segment: String(row.segment) as UserSegment,
          tone: String(row.tone),
          instructions: String(row.instructions),
          forbidden_words: this.forbiddenWords,
        },
      ]),
    );
    this.categories = [
      ...new Set(this.products.map((product) => product.category)),
    ].sort();
    this.#synonyms = this.#loadSynonyms(join(dataDir, "query_synonyms.json"));
  }

  close(): void {
    this.#database.close();
  }

  userProfile(userId: string, overrides: UserContext = {}): UserProfile {
    const base =
      this.profiles.get(userId) ??
      ({
        user_id: userId,
        segment: "new_user",
        preferred_categories: [],
        min_price_cents: 0,
        max_price_cents: 1_000_000,
        recent_views: [],
        recent_purchases: [],
      } satisfies UserProfile);
    return {
      ...base,
      preferred_categories: overrides.preferred_categories?.length
        ? overrides.preferred_categories
        : base.preferred_categories,
      // 单边价格约束代表本轮的新区间，另一端应开放，不能继承与它冲突的历史画像。
      min_price_cents:
        overrides.min_price_cents ??
        (overrides.max_price_cents === undefined ? base.min_price_cents : 0),
      max_price_cents:
        overrides.max_price_cents ??
        (overrides.min_price_cents === undefined
          ? base.max_price_cents
          : 1_000_000),
      recent_views: overrides.recent_views?.length
        ? overrides.recent_views
        : base.recent_views,
      recent_purchases: overrides.recent_purchases?.length
        ? overrides.recent_purchases
        : base.recent_purchases,
    };
  }

  search(input: {
    profile: UserProfile;
    categories: string[];
    min_price_cents: number;
    max_price_cents: number;
    tags: string[];
    limit: number;
  }): Product[] {
    if (
      input.min_price_cents < 0 ||
      input.max_price_cents <= 0 ||
      input.min_price_cents > input.max_price_cents
    ) {
      throw new CatalogError("invalid_product_search_price_range");
    }
    if (input.limit < 1 || input.limit > 20)
      throw new CatalogError("invalid_product_search_limit");
    const categories = normalize(input.categories);
    const tags = normalize(input.tags);
    if (input.categories.length && !categories.size)
      throw new CatalogError("invalid_product_search_categories");
    if (input.tags.length && !tags.size)
      throw new CatalogError("invalid_product_search_tags");

    return this.products
      .filter(
        (product) =>
          product.price_cents >= input.min_price_cents &&
          product.price_cents <= input.max_price_cents &&
          (!categories.size ||
            categories.has(product.category.toLocaleLowerCase())) &&
          (!tags.size ||
            product.tags.some((tag) => tags.has(tag.toLocaleLowerCase()))),
      )
      .sort(
        (left, right) =>
          this.#score(right, input.profile) - this.#score(left, input.profile),
      )
      .slice(0, input.limit);
  }

  inventory(productIds: string[]): Product[] {
    const products = new Map(
      this.#listProducts().map((product) => [product.product_id, product]),
    );
    return [...new Set(productIds)].flatMap((id) => {
      const product = products.get(id);
      return product && product.stock > 0 ? [product] : [];
    });
  }

  retrieveKnowledge(input: {
    query: string;
    categories: string[];
    product_ids: string[];
    limit: number;
  }): KnowledgeHit[] {
    if (input.limit < 1 || input.limit > 8)
      throw new CatalogError("invalid_knowledge_limit");
    // 类目是 Harness 已知的稳定检索词，优先于模型生成的长 query 进入 FTS token 上限。
    const expression = this.#matchExpression(
      this.#rewriteQuery([...input.categories, input.query].join(" ")),
    );
    if (!expression) return [];

    const filters: string[] = [];
    const parameters: Array<string | number> = [expression];
    if (input.categories.length) {
      filters.push(
        `f.category IN (${input.categories.map(() => "?").join(",")})`,
      );
      parameters.push(...input.categories);
    }
    if (input.product_ids.length) {
      filters.push(
        `(f.product_id IN (${input.product_ids.map(() => "?").join(",")}) OR f.product_id IS NULL)`,
      );
      parameters.push(...input.product_ids);
    }
    parameters.push(input.limit);
    const rows = this.#database.connection
      .prepare(
        `
      SELECT d.doc_id, d.title, f.chunk_ordinal, f.raw_content AS content,
             d.category, d.product_id, d.source, bm25(knowledge_documents_fts) AS rank
      FROM knowledge_documents_fts AS f JOIN knowledge_documents AS d ON d.doc_id = f.doc_id
      WHERE knowledge_documents_fts MATCH ? ${filters.length ? `AND ${filters.join(" AND ")}` : ""}
      ORDER BY rank, f.rowid LIMIT ?
    `,
      )
      .all(...parameters) as SqlRow[];
    return rows.map((row) => ({
      doc_id: String(row.doc_id),
      title: String(row.title),
      content: String(row.content),
      category: String(row.category),
      product_id: row.product_id === null ? null : String(row.product_id),
      source: String(row.source),
      chunk_ordinal: Number(row.chunk_ordinal),
      relevance_score:
        Math.round((1 / (1 + Math.abs(Number(row.rank)))) * 10_000) / 10_000,
    }));
  }

  marketingStrategy(segment: string): MarketingStrategy {
    const strategy = this.#templates.get(segment);
    if (!strategy) throw new CatalogError("unknown_marketing_segment");
    return strategy;
  }

  finalize(
    draft: RecommendationDraftItem[],
    request: RecommendationRequest,
    profile: UserProfile,
  ): RecommendedProduct[] {
    const current = new Map(
      this.#listProducts().map((product) => [product.product_id, product]),
    );
    const seen = new Set<string>();
    const recommendations: RecommendedProduct[] = [];
    for (const item of draft) {
      if (seen.has(item.product_id)) continue;
      const product = current.get(item.product_id);
      if (!product) throw new CatalogError("unknown_recommended_product");
      seen.add(item.product_id);
      if (
        product.stock <= 0 ||
        product.price_cents < profile.min_price_cents ||
        product.price_cents > profile.max_price_cents
      )
        continue;
      recommendations.push({
        product_id: product.product_id,
        name: product.name,
        category: product.category,
        price_cents: product.price_cents,
        brand: product.brand,
        stock: product.stock,
        tags: product.tags,
        score: this.#score(product, profile),
        low_stock: product.stock <= 100,
        reason: this.#sanitize(item.reason),
        marketing_copy: this.#sanitize(item.marketing_copy),
      });
      if (recommendations.length >= (request.num_items ?? 5)) break;
    }
    if (!recommendations.length)
      throw new CatalogError("no_available_recommendations");
    this.rememberPreference(
      request.user_id,
      request.context?.preferred_categories ?? [],
    );
    return recommendations;
  }

  rememberPreference(userId: string, categories: string[]): void {
    const profile = this.profiles.get(userId);
    const preferred = [
      ...new Set(categories.filter((category) => category.trim())),
    ];
    if (!profile || !preferred.length) return;
    this.#database.connection
      .prepare(
        "UPDATE user_profiles SET preferred_categories_json = ? WHERE user_id = ?",
      )
      .run(JSON.stringify(preferred), userId);
    this.profiles.set(userId, { ...profile, preferred_categories: preferred });
  }

  #listProducts(): Product[] {
    return (
      this.#database.connection
        .prepare("SELECT * FROM products ORDER BY product_id")
        .all() as SqlRow[]
    ).map(productFromRow);
  }

  #score(product: Product, profile: UserProfile): number {
    const preferred = normalize(profile.preferred_categories);
    const signals = normalize([
      ...profile.recent_views,
      ...profile.recent_purchases,
    ]);
    const searchable = normalize([
      product.name,
      product.category,
      ...product.tags,
    ]);
    let score = product.popularity_score * 0.55;
    if (preferred.has(product.category.toLocaleLowerCase())) score += 0.25;
    if ([...signals].some((signal) => searchable.has(signal))) score += 0.15;
    if (
      product.price_cents >= profile.min_price_cents &&
      product.price_cents <= profile.max_price_cents
    )
      score += 0.05;
    return Math.round(Math.min(score, 1) * 10_000) / 10_000;
  }

  #sanitize(text: string): string {
    return this.forbiddenWords.reduce(
      (result, word) => result.replaceAll(word, "***"),
      text,
    );
  }

  #loadSynonyms(path: string): Map<string, string[]> {
    const raw = JSON.parse(readFileSync(path, "utf8")) as Record<
      string,
      unknown
    >;
    const reversed = new Map<string, string[]>();
    for (const [canonical, variants] of Object.entries(raw)) {
      if (canonical.startsWith("_") || !Array.isArray(variants)) continue;
      for (const variant of variants) {
        if (typeof variant !== "string") continue;
        reversed.set(variant, [...(reversed.get(variant) ?? []), canonical]);
      }
    }
    return reversed;
  }

  #rewriteQuery(query: string): string {
    const extra = [...this.#synonyms.entries()]
      .filter(([variant]) => query.includes(variant))
      .flatMap(([, canonical]) => canonical)
      .filter(
        (term, index, terms) =>
          !query.includes(term) && terms.indexOf(term) === index,
      );
    return extra.length ? `${query} ${extra.join(" ")}` : query;
  }

  #matchExpression(query: string): string {
    return (query.toLocaleLowerCase().match(/[\p{L}\p{N}_]+/gu) ?? [])
      .slice(0, 8)
      .map(
        (token) => `"${segmentForIndex(token).trim().replaceAll('"', '""')}"`,
      )
      .join(" OR ");
  }
}
